modules/core/src/main/java/org/apache/ignite/internal/util/worker/GridWorker.java

```diff
@@ -289,6 +282,31 @@ protected void onIdle() {
             lsnr.onIdle(this);
     }
 
+    /**
+     * Callback on runner cancellation.
+     *
+     * @param first Whether the runner gets canceled the first time.
+     */
+    protected void onCancel(boolean first) {
```

Comment:
Don't like the name of the parameter. `firstCancelAttempt` or `firstCancelled` would be better.